File: python/21.py
# 合并两个有序链表
# Merge Two Sorted Lists

from common import ListNode, Optional, create_linked_list, print_linked_list


class Solution:
    def mergeTwoLists(
        self, list1: Optional[ListNode], list2: Optional[ListNode]
    ) -> Optional[ListNode]:

        dummy_head = ListNode()
        curr = dummy_head

        while list1 and list2:
            if list1.val <= list2.val:
                curr.next = list1
                list1 = list1.next
            else:
                curr.next = list2
                list2 = list2.next

            curr = curr.next

        curr.next = list1 if list1 else list2

        return dummy_head.next


if __name__ == "__main__":
    test_cases = [
        ([1, 2, 4], [1, 3, 4]),
        ([1, 2, 4], [])
    ]

    for test_case in test_cases:
        head1 = create_linked_list(test_case[0])
        head2 = create_linked_list(test_case[1])
        print_linked_list(head1)
        print_linked_list(head2)
        print_linked_list(Solution().mergeTwoLists(head1, head2))

File: rust/src/bin/21.rs
// 合并两个有序链表
// Merge Two Sorted Lists

use leetcode::common::{ListNode, create_linked_list, print_linked_list};

struct Solution {}

impl Solution {
    pub fn merge_two_lists(mut list1: Option<Box<ListNode>>, mut list2: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        let mut dummy_head = Some(Box::new(ListNode::new(0)));

        let mut curr = dummy_head.as_mut().unwrap();

        while list1.is_some() && list2.is_some() {
            if list1.as_ref().unwrap().val <= list2.as_ref().unwrap().val {
                curr.next = list1;
                list1 = curr.next.as_mut().unwrap().next.take();
            } else {
                curr.next = list2;
                list2 = curr.next.as_mut().unwrap().next.take();
            }
            curr = curr.next.as_mut().unwrap();
        }

        curr.next = list1.or(list2);

        dummy_head.unwrap().next
    }
}

fn main() {
    let arr1 = vec![1, 2, 4];
    let arr2 = vec![1, 3, 4];

    let list1 = create_linked_list(&arr1);
    let list2 = create_linked_list(&arr2);
    print_linked_list(&list1);
    print_linked_list(&list2);
    print_linked_list(&Solution::merge_two_lists(list1, list2));
}

